File: objc/Tests/TestSpectralIterBlocks.m
/*
 * TTI-O Objective-C Implementation
 * Copyright (c) 2026 The Thalion Initiative
 * SPDX-License-Identifier: LGPL-3.0-or-later
 *
 * -[TTIOAcquisitionRun iterBlocksFrom:to:threads:error:usingBlock:] and
 * the unit plan behind it.
 */
#import <Foundation/Foundation.h>
#import "Testing.h"
#import "Run/TTIOSpectralUnitPlan.h"

/* 6 spectra of 100 values; blocks start every 250 values, so no block
 * edge lands on a spectrum edge:
 *   spectra: [0,100) [100,200) [200,300) [300,400) [400,500) [500,600)
 *   blocks:  b0 [0,250)   b1 [250,500)   b2 [500,...)
 *   owner by first value: s0,s1,s2 -> b0 ; s3,s4 -> b1 ; s5 -> b2
 */
static void siuPlanBasics(void)
{
    unsigned long long off[6] = {0, 100, 200, 300, 400, 500};
    unsigned int len[6]       = {100, 100, 100, 100, 100, 100};
    unsigned long long bs[3]  = {0, 250, 500};

    NSArray<NSValue *> *u = [TTIOSpectralUnitPlan unitsForSpectraFrom:0 to:6
                                                             offsets:off lengths:len
                                                    blockValueStarts:bs count:3];
    PASS(u.count == 3, "unit plan: 3 units for 6 spectra over 3 blocks (%lu)",
         (unsigned long)u.count);
    if (u.count != 3) return;

    TTIOSpectralUnit a; [u[0] getValue:&a];
    PASS(a.block == 0 && a.firstSpectrum == 0 && a.nSpectra == 3,
         "unit plan: b0 owns s0..s2 (block=%lu first=%lu n=%lu)",
         (unsigned long)a.block, (unsigned long)a.firstSpectrum,
         (unsigned long)a.nSpectra);
    PASS(a.valueStart == 0 && a.valueEnd == 300,
         "unit plan: b0 extent [0,300) (%llu,%llu)", a.valueStart, a.valueEnd);

    TTIOSpectralUnit b; [u[1] getValue:&b];
    PASS(b.block == 1 && b.firstSpectrum == 3 && b.nSpectra == 2,
         "unit plan: b1 owns s3..s4 (block=%lu first=%lu n=%lu)",
         (unsigned long)b.block, (unsigned long)b.firstSpectrum,
         (unsigned long)b.nSpectra);
    /* s4 ends at 500, which is b2's first value: the extent runs past
     * the owning block's end, so units are not byte independent. */
    PASS(b.valueStart == 300 && b.valueEnd == 500,
         "unit plan: b1 extent [300,500) crosses its own block end (%llu,%llu)",
         b.valueStart, b.valueEnd);

    TTIOSpectralUnit c; [u[2] getValue:&c];
    PASS(c.block == 2 && c.firstSpectrum == 5 && c.nSpectra == 1,
         "unit plan: b2 owns s5 (block=%lu first=%lu n=%lu)",
         (unsigned long)c.block, (unsigned long)c.firstSpectrum,
         (unsigned long)c.nSpectra);
}

/* A range starting part-way into a block reports run-global indices. */
static void siuPlanMidBlockStart(void)
{
    unsigned long long off[6] = {0, 100, 200, 300, 400, 500};
    unsigned int len[6]       = {100, 100, 100, 100, 100, 100};
    unsigned long long bs[3]  = {0, 250, 500};

    NSArray<NSValue *> *u = [TTIOSpectralUnitPlan unitsForSpectraFrom:1 to:4
                                                             offsets:off lengths:len
                                                    blockValueStarts:bs count:3];
    PASS(u.count == 2, "unit plan: mid-block range spans 2 units (%lu)",
         (unsigned long)u.count);
    if (u.count != 2) return;

    TTIOSpectralUnit a; [u[0] getValue:&a];
    PASS(a.firstSpectrum == 1 && a.nSpectra == 2,
         "unit plan: clipped head unit is s1..s2 (first=%lu n=%lu)",
         (unsigned long)a.firstSpectrum, (unsigned long)a.nSpectra);
    PASS(a.valueStart == 100 && a.valueEnd == 300,
         "unit plan: clipped head extent [100,300) (%llu,%llu)",
         a.valueStart, a.valueEnd);

    TTIOSpectralUnit b; [u[1] getValue:&b];
    PASS(b.firstSpectrum == 3 && b.nSpectra == 1,
         "unit plan: clipped tail unit is s3 only (first=%lu n=%lu)",
         (unsigned long)b.firstSpectrum, (unsigned long)b.nSpectra);
}

/* One spectrum longer than a block: the blocks it spans own no spectrum
 * start and must not produce empty units. */
static void siuPlanSpectrumLongerThanBlock(void)
{
    unsigned long long off[2] = {0, 1000};
    unsigned int len[2]       = {1000, 10};
    unsigned long long bs[5]  = {0, 250, 500, 750, 1000};

    NSArray<NSValue *> *u = [TTIOSpectralUnitPlan unitsForSpectraFrom:0 to:2
                                                             offsets:off lengths:len
                                                    blockValueStarts:bs count:5];
    PASS(u.count == 2, "unit plan: straddling spectrum yields 2 units, not 5 (%lu)",
         (unsigned long)u.count);
    BOOL noneEmpty = YES;
    for (NSValue *v in u) {
        TTIOSpectralUnit s; [v getValue:&s];
        if (s.nSpectra < 1) noneEmpty = NO;
    }
    PASS(noneEmpty, "unit plan: no empty unit emitted");
}

static void siuPlanEmptyRange(void)
{
    unsigned long long off[2] = {0, 100};
    unsigned int len[2]       = {100, 100};
    unsigned long long bs[1]  = {0};
    NSArray<NSValue *> *u = [TTIOSpectralUnitPlan unitsForSpectraFrom:3 to:3
                                                             offsets:off lengths:len
                                                    blockValueStarts:bs count:1];
    PASS(u.count == 0, "unit plan: empty range yields no units (%lu)",
         (unsigned long)u.count);
}

/* Every spectrum exactly once, no gaps, no overlap. */
static void siuPlanPartitions(void)
{
    const NSUInteger N = 97;
    unsigned long long off[97]; unsigned int len[97];
    unsigned long long cursor = 0;
    for (NSUInteger i = 0; i < N; i++) {
        len[i] = (unsigned int)(7 + (i % 13));
        off[i] = cursor; cursor += len[i];
    }
    unsigned long long bs[9];
    for (NSUInteger b = 0; b < 9; b++) bs[b] = b * (cursor / 9);

    NSArray<NSValue *> *u = [TTIOSpectralUnitPlan unitsForSpectraFrom:0 to:N
                                                             offsets:off lengths:len
                                                    blockValueStarts:bs count:9];
    NSMutableIndexSet *seen = [NSMutableIndexSet indexSet];
    NSUInteger expectNext = 0;
    BOOL tiles = YES;
    for (NSValue *v in u) {
        TTIOSpectralUnit s; [v getValue:&s];
        if (s.firstSpectrum != expectNext) tiles = NO;
        [seen addIndexesInRange:NSMakeRange(s.firstSpectrum, s.nSpectra)];
        expectNext = s.firstSpectrum + s.nSpectra;
    }
    PASS(tiles, "unit plan: units tile without gaps or overlap");
    PASS(seen.count == N && expectNext == N,
         "unit plan: all %lu spectra covered exactly once (%lu)",
         (unsigned long)N, (unsigned long)seen.count);
}

void testSpectralIterBlocks(void)
{
    siuPlanBasics();
    siuPlanMidBlockStart();
    siuPlanSpectrumLongerThanBlock();
    siuPlanEmptyRange();
    siuPlanPartitions();
}

File: python/tests/test_spectral_iter_blocks.py
"""``AcquisitionRun.for_each_block``: the spectral parallel block consumer.

Parity with Objective-C ``-iterBlocksFrom:to:threads:error:usingBlock:``
and Java ``AcquisitionRun.iterBlocks``.

SPDX-License-Identifier: Apache-2.0
"""
from __future__ import annotations

import threading
from pathlib import Path

import numpy as np
import pytest

from ttio import SpectralDataset, WrittenRun
from ttio.enums import AcquisitionMode

# The FDZ1 block is 2**20 values and is not configurable, so the corpus
# is sized past it on purpose: 600 spectra of 2000 points is 1200000
# values per channel, which is 2 blocks. A smaller corpus is a single
# block and never crosses a unit boundary.
N_SPEC = 600
N_PTS = 2000


def _make_run(codec: str | None = None) -> WrittenRun:
    """m/z value j of spectrum i is 1000*i + j, so a spectrum's content
    names the spectrum and no assertion needs a derived index."""
    offsets = np.arange(N_SPEC, dtype=np.uint64) * N_PTS
    lengths = np.full(N_SPEC, N_PTS, dtype=np.uint32)
    mz = np.concatenate([1000.0 * i + np.arange(N_PTS, dtype=np.float64)
                         for i in range(N_SPEC)])
    intensity = np.concatenate([((np.arange(N_PTS, dtype=np.float64) + 7 * i) % 977)
                                for i in range(N_SPEC)])
    kw = {"signal_compression": codec} if codec else {}
    return WrittenRun(
        spectrum_class="TTIOMassSpectrum",
        acquisition_mode=int(AcquisitionMode.MS1_DDA),
        channel_data={"mz": mz, "intensity": intensity},
        offsets=offsets,
        lengths=lengths,
        retention_times=np.linspace(0.0, float(N_SPEC), N_SPEC, dtype=np.float64),
        ms_levels=np.ones(N_SPEC, dtype=np.int32),
        polarities=np.ones(N_SPEC, dtype=np.int32),
        precursor_mzs=np.zeros(N_SPEC, dtype=np.float64),
        precursor_charges=np.zeros(N_SPEC, dtype=np.int32),
        base_peak_intensities=np.full(N_SPEC, 100.0, dtype=np.float64),
        **kw)


def _write(tmp_path_factory, name: str, codec: str | None) -> Path:
    out = tmp_path_factory.mktemp(name) / f"{name}.tio"
    SpectralDataset.write_minimal(
        out, title=name, isa_investigation_id=f"TTIO:{name}",
        runs={"run_0001": _make_run(codec)})
    return out


@pytest.fixture(scope="module")
def corpus(tmp_path_factory) -> Path:
    """Codec 17 channels, so units come from the FDZ1 block tables."""
    return _write(tmp_path_factory, "fdz", None)


@pytest.fixture(scope="module")
def batched_corpus(tmp_path_factory) -> Path:
    """Numpress channels have no FDZ1 stream, so units are spectrum
    counts."""
    return _write(tmp_path_factory, "npz", "numpress_delta")


def _collect(run, threads=None, start=0, stop=None):
    """m/z[0] per run index, gathered through for_each_block."""
    got: dict[int, float] = {}
    units: list[int] = []
    lock = threading.Lock()

    def fn(view, view_start, first_spectrum, n_spectra):
        local = {}
        for k in range(n_spectra):
            sp = view[view_start + k]
            local[first_spectrum + k] = float(np.asarray(sp.mz_array.data)[0])
        with lock:
            got.update(local)
            units.append(n_spectra)

    run.for_each_block(fn, start, stop, threads=threads)
    return got, units


def test_the_corpus_spans_several_units(corpus, batched_corpus):
    """Without this the suite is a false green: a single-unit corpus
    passes every assertion below without ever crossing a boundary."""
    with SpectralDataset.open(corpus) as ds:
        run = ds.ms_runs["run_0001"]
        assert run._fdz_channels, "the codec-17 corpus lost its FDZ1 channels"
        assert len(run._spectral_units(0, N_SPEC)) >= 2
    with SpectralDataset.open(batched_corpus) as ds:
        run = ds.ms_runs["run_0001"]
        assert not run._fdz_channels, "the numpress corpus kept an FDZ1 channel"
        assert len(run._spectral_units(0, N_SPEC)) >= 2


@pytest.mark.parametrize("threads", [None, 1, 2, 4, 8])
def test_every_spectrum_carries_its_own_content(corpus, threads):
    with SpectralDataset.open(corpus) as ds:
        got, units = _collect(ds.ms_runs["run_0001"], threads=threads)
    assert len(got) == N_SPEC, f"visited {len(got)} of {N_SPEC} in {len(units)} units"
    for i in range(N_SPEC):
        assert got[i] == pytest.approx(1000.0 * i), f"spectrum {i} carried {got[i]}"


def test_units_tile_the_range_without_gap_or_overlap(corpus):
    with SpectralDataset.open(corpus) as ds:
        units = ds.ms_runs["run_0001"]._spectral_units(0, N_SPEC)
    expect = 0
    for _blk, first, count, vstart, vend in units:
        assert first == expect, f"gap or overlap at {expect}"
        assert count >= 1, "empty unit emitted"
        assert vend > vstart
        expect = first + count
    assert expect == N_SPEC


def test_a_range_starting_part_way_in_reports_run_global_indices(corpus):
    """The shape that returned the wrong records on the genomic side."""
    lo, hi = 37, 461
    with SpectralDataset.open(corpus) as ds:
        got, _ = _collect(ds.ms_runs["run_0001"], threads=2, start=lo, stop=hi)
    assert sorted(got) == list(range(lo, hi))
    for i in range(lo, hi):
        assert got[i] == pytest.approx(1000.0 * i)


def test_matches_iter_spectra(corpus):
    with SpectralDataset.open(corpus) as ds:
        run = ds.ms_runs["run_0001"]
        ordered = [float(np.asarray(sp.mz_array.data)[0]) for sp in run.iter_spectra()]
        got, _ = _collect(run, threads=4)
    assert ordered == [got[i] for i in range(N_SPEC)]


def test_spectrum_count_batches_deliver_every_spectrum(batched_corpus):
    """The tier used when no channel carries an FDZ1 stream."""
    with SpectralDataset.open(batched_corpus) as ds:
        got, _ = _collect(ds.ms_runs["run_0001"], threads=4)
    assert len(got) == N_SPEC
    for i in range(N_SPEC):
        # Numpress is fixed point, so compare within its quantisation.
        assert got[i] == pytest.approx(1000.0 * i, rel=1e-6)


def test_empty_range_visits_nothing(corpus):
    calls = []
    with SpectralDataset.open(corpus) as ds:
        ds.ms_runs["run_0001"].for_each_block(lambda *a: calls.append(a), 5, 5)
    assert calls == []
